### PR TITLE
InfiniteScroll: fix avoid event trigger when element invisible

### PR DESCRIPTION
**Pull request by qingdengyue**
Please make sure these boxes are checked before submitting your PR, thank you!

* [x] Make sure you follow Element's contributing guide ([中文](https://github.com/ElemeFE/element/blob/master/.github/CONTRIBUTING.zh-CN.md) | [English](https://github.com/ElemeFE/element/blob/master/.github/CONTRIBUTING.en-US.md) | [Español](https://github.com/ElemeFE/element/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/ElemeFE/element/blob/master/.github/CONTRIBUTING.fr-FR.md)).
* [x] Make sure you are merging your commits to `dev` branch.
* [x] Add some descriptions and refer relative issues for you PR.

### CODE REVIEW

**Comment by qingdengyue** [comment]
#16673

**Comment by a631807682** [comment]
`scroll` event does not trigger after component visible.
When contents don't fill the container at first, `InfiniteScroll` can not do anything.

**Comment by qingdengyue** [comment]
> `scroll` event does not trigger after component visible.
> When contents don't fill the container at first, `InfiniteScroll` can not do anything.
> I think when component hidden, it should create a observer to watch it if visible, and then emit 'scroll' event to compute should trigger callback or not.

yeah. when the element is invisible,may be its parent element or self invisible. now when trigger scroll event,it will check the element visible status,if true ,go on ,if  false,just return. the scroll event like a observer.

**Comment by a631807682** [comment]
> > `scroll` event does not trigger after component visible.
> > When contents don't fill the container at first, `InfiniteScroll` can not do anything.
> > I think when component hidden, it should create a observer to watch it if visible, and then emit 'scroll' event to compute should trigger callback or not.
> 
> yeah. when the element is invisible,may be its parent element or self invisible. now when trigger scroll event,it will check the element visible status,if true ,go on ,if false,just return. the scroll event like a observer.

yeah, but the problem is no scroll at first.
when component invisible, `load` will not trigger, and `count` equal `0`. and then component invisible, but `count` still equal `0`, there are no `scroll` and we have no way to trgger it. unless dispatch event
like this `this.$refs.table.$el.dispatchEvent(new CustomEvent('scroll'))`.

template
```
...
<ul v-infinite-scroll="load" style="height: 300px;overflow: auto;">
     <li v-for="i in count" style="display: flex;height: 50px;">{{ i }}</li>
</ul>
...
```
js
```
...
data() {
  return {
    count: 0
  }
}
...
```

**Comment by qingdengyue** [comment]
> > > `scroll` event does not trigger after component visible.
> > > When contents don't fill the container at first, `InfiniteScroll` can not do anything.
> > > I think when component hidden, it should create a observer to watch it if visible, and then emit 'scroll' event to compute should trigger callback or not.
> > 
> > 
> > yeah. when the element is invisible,may be its parent element or self invisible. now when trigger scroll event,it will check the element visible status,if true ,go on ,if false,just return. the scroll event like a observer.
> 
> yeah, but the problem is no scroll at first.
> when component invisible, `load` will not trigger, and `count` equal `0`. and then component visible, but `count` still equal `0`, there are no `scroll` and we have no way to trgger it. unless dispatch event
> like this `this.$refs.table.$el.dispatchEvent(new CustomEvent('scroll'))`.
> 
> template
> 
> ```
> ...
> <ul v-infinite-scroll="load" style="height: 300px;overflow: auto;">
>      <li v-for="i in count" style="display: flex;height: 50px;">{{ i }}</li>
> </ul>
> ...
> ```
> 
> js
> 
> ```
> ...
> data() {
>   return {
>     count: 0
>   }
> }
> ...
> ```

v-infinite-scroll directive will bind the container's  scroll event on inserted hooks. so if word is not correct.you can provider a reproduction link?

**Comment by a631807682** [comment]
link
---
https://codepen.io/a631807682/pen/jgqqJy?editors=1011#0

i chang the link to `//deploy-preview-16681--element.netlify.com/element-ui.4b9a2bf.js` , is a way to provider a reproduction link?

setp
---
1. click `change visible` button
2. can not scroll ( contents don't fill the container what `infinite-scroll-immediate` do)

@qingdengyue

**Comment by iamkun** [comment]
> ## link
> https://codepen.io/a631807682/pen/jgqqJy?editors=1011#0
> 
> i chang the link to `//deploy-preview-16681--element.netlify.com/element-ui.4b9a2bf.js` , is a way to provider a reproduction link?
> 
> ## setp
> 1. click `change visible` button
> 2. can not scroll ( contents don't fill the container what `infinite-scroll-immediate` do)
> 
> @qingdengyue

@qingdengyue same problem to me, please check the link above

File: packages/infinite-scroll/src/main.js [hunk 1]
```diff
@@ -84,13 +84,20 @@ const getScrollOptions = (el, vm) => {
...
+  let rect = getElementContainer(el).getBoundingClientRect();
+  return rect.width > 0 && rect.height > 0;
+};
 
 const handleScroll = function(cb) {
   const { el, vm, container, observer } = this[scope];
   const { distance, disabled } = getScrollOptions(el, vm);
 
-  if (disabled) return;
+  if (disabled || !isElementVisible(el)) return;
```
Comment (a631807682) [inline packages/infinite-scroll/src/main.js]: i suggest to create `IntersectionObserver` to watch element visible status when invisible, like this:
```
if (!isElementVisible(el) && !visibleObserver) {
	visibleObserver = new IntersectionObserver(onScroll);
	visibleObserver.observe(el);
}
```

**Pull request closed without merging**